runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+		err = os.RemoveAll(filepath.Join(d.readPath, entry.Name(), readVersion))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *db) Size() int64 {
+	var paths []string
+	entries, err := os.ReadDir(d.readPath)
+	if err != nil { // ignore error
+		return 0
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		// this is to avoid counting temp tables during source ingestion
+		// in certain cases we only want to compute the size of the serving db files
+		// TODO :: remove this when removing staged table concepts
+		if strings.HasPrefix(entry.Name(), "__rill_tmp_") {
+			continue
+		}
+		path := filepath.Join(d.readPath, entry.Name())
+		version, exist, _ := d.readTableVersion(entry.Name())
+		if !exist {
+			continue
+		}
+		paths = append(paths, filepath.Join(path, fmt.Sprintf("%s.db", version)))
+	}
+	return fileSize(paths)
+}
+
+// acquireWriteConn syncs the write database, initializes the write handle and returns a write connection.
+// The release function should be called to release the connection.
+// It should be called with the writeMu locked.
+func (d *db) acquireWriteConn(ctx context.Context) (*sqlx.Conn, func() error, error) {
+	err := d.syncWrite(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	db, err := d.openDBAndAttach(ctx, false)
+	if err != nil {
+		return nil, nil, err
+	}
+	conn, err := db.Connx(ctx)
+	if err != nil {
+		_ = db.Close()
+		return nil, nil, err
+	}
+	return conn, func() error {
+		_ = conn.Close()
+		err = db.Close()
+		return err
+	}, nil
+}
+
+func (d *db) openDBAndAttach(ctx context.Context, read bool) (*sqlx.DB, error) {
+	// open the db
+	var (
+		dsn      *url.URL
+		err      error
+		settings map[string]string
+		path     string
+	)
+	if read {
+		dsn, err = url.Parse("") // in-memory
+		settings = d.opts.ReadSettings
+		path = d.readPath
+	} else {
+		path = d.writePath
+		dsn, err = url.Parse(filepath.Join(path, "stage.db"))
+		settings = d.opts.WriteSettings
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	query := dsn.Query()
+	for k, v := range settings {
+		query.Set(k, v)
+	}
+	dsn.RawQuery = query.Encode()
+	connector, err := duckdb.NewConnector(dsn.String(), func(execer driver.ExecerContext) error {
+		for _, qry := range d.opts.InitQueries {
+			_, err := execer.ExecContext(context.Background(), qry, nil)
+			if err != nil && strings.Contains(err.Error(), "Failed to download extension") {
+				// Retry using another mirror. Based on: https://github.com/duckdb/duckdb/issues/9378
+				_, err = execer.ExecContext(context.Background(), qry+" FROM 'http://nightly-extensions.duckdb.org'", nil)
+			}
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	db := sqlx.NewDb(otelsql.OpenDB(connector), "duckdb")
+
+	err = otelsql.RegisterDBStatsMetrics(db.DB, otelsql.WithAttributes(attribute.String("db.system", "duckdb"), attribute.String("db_identifier", d.dbIdentifier)))
```

Comment:
It would be more useful to have `instance_id` and maybe the connector name here – maybe these attributes should be provided from upstream?